src/NuGet.Clients/NuGet.VisualStudio.Common/Experimentation/NuGetExperimentationService.cs

```diff
@@ -14,32 +15,51 @@ public class NuGetExperimentationService : INuGetExperimentationService
     {
         private readonly IEnvironmentVariableReader _environmentVariableReader;
         private readonly IExperimentationService _experimentationService;
+        private readonly Lazy<IOutputConsoleProvider> _outputConsoleProvider;
 
-        public NuGetExperimentationService()
-            : this(EnvironmentVariableWrapper.Instance, ExperimentationService.Default)
+        [ImportingConstructor]
+        public NuGetExperimentationService(Lazy<IOutputConsoleProvider> _outputConsoleProvider)
+            : this(EnvironmentVariableWrapper.Instance, ExperimentationService.Default, _outputConsoleProvider)
         {
             // ensure uniqueness.
         }
 
-        internal NuGetExperimentationService(IEnvironmentVariableReader environmentVariableReader, IExperimentationService experimentationService)
+        internal NuGetExperimentationService(IEnvironmentVariableReader environmentVariableReader, IExperimentationService experimentationService, Lazy<IOutputConsoleProvider> outputConsoleProvider)
         {
             _environmentVariableReader = environmentVariableReader ?? throw new ArgumentNullException(nameof(environmentVariableReader));
             _experimentationService = experimentationService ?? throw new ArgumentNullException(nameof(experimentationService));
+            _outputConsoleProvider = outputConsoleProvider ?? throw new ArgumentNullException(nameof(outputConsoleProvider));
         }
 
         public bool IsExperimentEnabled(ExperimentationConstants experiment)
         {
             var isExpForcedEnabled = false;
             var isExpForcedDisabled = false;
-            if (!string.IsNullOrEmpty(experiment.FlightEnvironmentVariable))
+            string flightVariable = experiment.FlightEnvironmentVariable;
+
+            if (!string.IsNullOrEmpty(flightVariable))
             {
-                string envVarOverride = _environmentVariableReader.GetEnvironmentVariable(experiment.FlightEnvironmentVariable);
+                string envVarOverride = _environmentVariableReader.GetEnvironmentVariable(flightVariable);
 
                 isExpForcedDisabled = envVarOverride == "0";
                 isExpForcedEnabled = envVarOverride == "1";
+
+                if (isExpForcedDisabled || isExpForcedEnabled)
+                {
+                    LogEnvironmentVariableOverride(experiment.FlightFlag, flightVariable, envVarOverride);
+                }
             }
 
             return !isExpForcedDisabled && (isExpForcedEnabled || _experimentationService.IsCachedFlightEnabled(experiment.FlightFlag));
         }
+
+        private void LogEnvironmentVariableOverride(string flightName, string variableName, string variableValue)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
```

Comment:
Should the logging be `cached` to ensure it's done only once, or are we ok with logging every time the check is done?

Ofc the trick with caching is that it's theoretically possible for the value to change mid process.

Reply:
VS caches the environment variable evaluation when the process starts so the env vars can't change after VS starts. However, I could cache the message logging but if the user clears the Output and then wants to verify the override they'd have to restart VS. We could have a mechanism to clear the cached output, but I think that might be overkill.

Since this is an override that's intended for limited use, debugging, and testing, I'd err on the side of logging every occurrence of the override. Does that sound good to you? Happy to hear your thoughts if you think otherwise!

Reply:
Given that it won't affect anyone by default, or frequently, I think it's a fair approach, and I fully trust your judgment here.